proposals/records/1-index-scaling.md

```diff
@@ -0,0 +1,183 @@
+- Long term NucliaDB normal "usage"
+- How pricing changes over time affects usage
+- Proper shard size
+
+
+## Definitions
+
+- Shard: the storage of a knowledge box's indexes(keyword, bm25, graph, vector).
+  We currently store multiple replicas of a knowledge box's indexes across different nodes.
+- Node: Storage, writer, reader and sidecar(currently) for a set of shards that
+  contain index data. This is separate from resource data.
+- IndexNode: For the purpose of this proposal and to prevent confusion around naming,
+  a Node will be called `IndexNode`
+- IndexNodeCoordinator: Layer that manages IndexNode searches, scaling and load balancing
+
+
+## Current solution/situation
+
+Currently, an IndexNode is deployed for writes(indexing) and reads(searching) for a set
+of shards. Each shard is replicated and read(search) requests are typically
+round robin load balanced against the IndexNodes that hold replicas.
+
+Scaling this design requires the following:
+- indexing replicas needs to be known at write time
+- no current way to increase replica count on existing data
+- scale disk: physically copy/move shards around between IndexNodes to rebalance
+- disk and number of shards per node is the only way to determine a node is "full"
+- (number of shards is a poor metric to determine is an IndexNode is full)
+- each indexing operation happens on each shard replica a resource is on
+
+Finally, search requests are round robin load balanced against the IndexNodes a knowledge box
+shard replicas are located on(2 shard replicas typically) from the search component:
+
+```mermaid
+flowchart TD
+    Search --> KBIndexNodeReplica1
+    Search --> KBIndexNodeReplica2
+```
+
+
+## Proposed Solution
+
+Create a new IndexNodeCoordinator component.
+
+The responsibilies of this component will be:
+- Manage IndexNodes and the shard replicas they contain
+  - This information will not be managed by ingest anymore and ingest will no longer even need to know about NODES
+- Index Consumer: Index data against all shard replicas and validate consistency
+- Search proxy: Search requests go to this component, which knows about all
+  IndexNodes and which one is best to perform search against
+- Increase/descrease replica count by individual shard and move shards around depending on IndexNode load
+- Manage shard reindexing / shadow shard implementation 
+
+This proposal focuses on an implementation that will cause minimal impact on the existing
+architecture and allows us to test/gather feedback on feasability quickly.
+
+### IndexNode Shard replication:
+
+The IndexNodeCoordinator will ask 1 IndexNode to copy it's shard data to another IndexNode:
+
+
+```mermaid
+flowchart TB
+    IndexNodeCoordinator-- CopyShardRequest -->SourceIndexNode
+    SourceIndexNode-- CopyShard -->DestinationIndexNode
+    IndexNodeCoordinator-- ValidateShard -->DestinationIndexNode
+```
+
+When an IndexNode needs it's replicas increased:
+
+- All indexing to shard needs to be buffered
```

Comment:
do you think we can use shadow shards for buffering?

Reply:
but we still need to finish playing all the data onto the shards that are being worked on so I'm not sure I see how shadow shard will help here.